parser: Adjust parser to accept integer hex literals

zummenix opened the pull request:
#385, This supports only integer hex literals.

zummenix commented on parser/src/token.rs:
```diff
@@ -752,6 +825,56 @@ mod test {
+                (r#"       ~~~~~           "#, IntLiteral(291)),
+                (r#"            ~~~~~      "#, IntLiteral(1)),
+                (r#"                 ~~~~  "#, IntLiteral(-10)),
+                (r#"                     ~~"#, IntLiteral(0)),
```
Should we forbid this?

Marwes replied:
I'd say yes, it seems weird since it is sort of a prefix followed by nothing. Tried it in rust and it complained abot "a number without any digits aswell" so that is some precedence for rejecting it.

Marwes commented on parser/src/token.rs:
```diff
@@ -516,6 +548,47 @@ impl<'input> Iterator for Tokenizer<'input> {
+///
+/// This is basically a copy and adaptation of `std::num::from_str_radix`.
+fn i64_from_hex(hex: &str, is_positive: bool) -> Result<i64, Error> {
+    let radix = 16;
```
Nit: Could be a `const`

Marwes replied:
Nit: Could be a const

Marwes commented on parser/src/token.rs:
```diff
@@ -516,6 +548,47 @@ impl<'input> Iterator for Tokenizer<'input> {
+    let mut result = 0i64;
+    if is_positive {
+        for &c in digits {
+            let x = match (c as char).to_digit(radix) {
```
The string is already validated to be all hex digits at this point so just do `expect` here which also makes it possible to remove the `HexLiteralInvalid` case.

Marwes commented on parser/src/token.rs:
```diff
@@ -516,6 +548,47 @@ impl<'input> Iterator for Tokenizer<'input> {
+                Some(result) => result,
+                None => return Err(HexLiteralUnderflow),
+            };
+        }
```
I assume you split the positive and negative case to deal with abs(i64::MIN) != abs(i64::MAX) but I think it should also work if `x` is multiplied by the sign before each `checked_add`. What do you think? It should remove the duplication I think.

Marwes commented on parser/src/token.rs:
```diff
@@ -516,6 +548,47 @@ impl<'input> Iterator for Tokenizer<'input> {
+                Some(x) => x,
+                None => return Err(HexLiteralInvalid),
+            };
+            result = match result.checked_mul(radix as i64) {
```
nit: Could write the option matches as `result.checked_mul(radix as i64).ok_or(HexLiteralInvalid)?`

Marwes commented on parser/src/token.rs:
```diff
@@ -412,6 +428,22 @@ impl<'input> Tokenizer<'input> {
             }
+            Some((end, 'x')) => {
+                self.bump(); // Skip 'x'
+                let (end, hex) = self.take_while(end.shift('x'), is_hex);
```
This needs a check that the next character is not an identifier otherwise `0x1q` will be tokenized as `0x1 q` . This is done later in the function for normal (decimal) integers but is actually also bugged for floats! That is, `1.0q` is accepted today. If you want it would be great if you can add the same check for floats as well  😅

Marwes commented on parser/src/token.rs:
```diff
@@ -412,6 +428,22 @@ impl<'input> Tokenizer<'input> {
+                        }
+                    }
+                    _ => {
+                        return  self.error(start, HexLiteralWrongPrefix);
```
super nit: Extra whitespace after return (just run it through rustfmt if you have it).

zummenix commented:
I think I addressed all issues.
Also fixed float and byte literals with unexpected char.

P.S: I noticed that checking of the location in tests doesn't work. For example in:
```rust
assert_eq!(
    tokenizer(r#"0x1q"#).last(),
    Some(error(loc(3), UnexpectedChar('q')))
);
```
it doesn't matter if `loc(3)` or `loc(0)` test succeeds anyway. I'm not sure this is intentional.

Marwes commented:
👍 

> it doesn't matter if loc(3) or loc(0) test succeeds anyway. I'm not sure this is intentional.

That is actually intentional. There are a bunch of [parser tests](https://github.com/gluon-lang/gluon/blob/22863d39034e8ffd56e929e7a0258f544f3fe1bd/parser/tests/basic.rs#L711-L745) which compares the AST produced by the parser to one that is manually created and I didn't want to spec out the spans for each individual expression in those (since it is relevant for those tests and its quite difficult to mess up the spans so it is not worthwhile to test).

In hindsight though it would be better to traverse the AST generated by the parser and zero out all spans before doing those comparisons though. Then spans could matter in `PartialEq for Span`.

Marwes commented:
Thanks again for the PR!
